File: KNIRVORACLE_NEW/pkg/network/tunnelclient.go
package network

import (
	"bufio"
	"encoding/json"
	"fmt"
	"log"
	"net"
	"strings"
	"sync"
	"time"

	"KNIRVORACLE/config"
)

// TunnelClientImpl manages the connection to a central relay server for NAT traversal
type TunnelClientImpl struct {
	config          *config.TunnelClientConfig
	nodeID          string
	chainID         string
	internalIP      string
	internalP2PPort uint
	nodeType        string
	conn            net.Conn
	isConnected     bool
	reconnectCount  int
	mutex           sync.Mutex
	stopChan        chan struct{}
	pingTicker      *time.Ticker
}

// IdentifyMessage is sent to the tunnel server to identify this node
type IdentifyMessage struct {
	Action          string `json:"action"`
	PeerID          string `json:"devId"`
	ChainID         string `json:"chainId"`
	InternalIP      string `json:"internalIp"`
	InternalP2PPort uint   `json:"internalP2pPort"`
	Type            string `json:"type"`
}

// PingMessage is sent periodically to keep the connection alive
type PingMessage struct {
	Action string `json:"action"`
}

// ServerMessage represents messages received from the tunnel server
type ServerMessage struct {
	Status  string `json:"status,omitempty"`
	Message string `json:"message,omitempty"`
	Action  string `json:"action,omitempty"`
}

// NewTunnelClient creates a new tunnel client
func NewTunnelClient(cfg *config.TunnelClientConfig, nodeID, chainID, internalIP string, internalP2PPort uint, nodeType string) *TunnelClientImpl {
	return &TunnelClientImpl{
		config:          cfg,
		nodeID:          nodeID,
		chainID:         chainID,
		internalIP:      internalIP,
		internalP2PPort: internalP2PPort,
		nodeType:        nodeType,
		isConnected:     false,
		reconnectCount:  0,
		stopChan:        make(chan struct{}),
	}
}

// Connect establishes a connection to the tunnel server
func (c *TunnelClientImpl) Connect() error {
	c.mutex.Lock()
	defer c.mutex.Unlock()

	if c.isConnected {
		return nil // Already connected
	}

	// Format the address properly for both IPv4 and IPv6
	var serverAddr string
	if strings.Contains(c.config.ServerAddress, ":") {
		// IPv6 address needs to be enclosed in square brackets
		serverAddr = fmt.Sprintf("[%s]:%d", c.config.ServerAddress, c.config.ControlPort)
	} else {
		// IPv4 address
		serverAddr = fmt.Sprintf("%s:%d", c.config.ServerAddress, c.config.ControlPort)
	}
	log.Printf("Connecting to tunnel server at %s...", serverAddr)

	var err error
	c.conn, err = net.Dial("tcp", serverAddr)
	if err != nil {
		return fmt.Errorf("failed to connect to tunnel server: %w", err)
	}

	// Send identification message
	identifyMsg := IdentifyMessage{
		Action:          "IDENTIFY",
		PeerID:          c.nodeID,
		ChainID:         c.chainID,
		InternalIP:      c.internalIP,
		InternalP2PPort: c.internalP2PPort,
		Type:            c.nodeType,
	}

	identifyBytes, err := json.Marshal(identifyMsg)
	if err != nil {
		c.conn.Close()
		return fmt.Errorf("failed to marshal identify message: %w", err)
	}

	_, err = c.conn.Write(append(identifyBytes, '\n'))
	if err != nil {
		c.conn.Close()
		return fmt.Errorf("failed to send identify message: %w", err)
	}

	// Start reading responses in a goroutine
	go c.readResponses()

	c.isConnected = true
	c.reconnectCount = 0
	log.Printf("Connected to tunnel server and sent identification")

	// Start ping ticker
	c.startPingTicker()

	return nil
}

// readResponses reads and processes messages from the tunnel server
func (c *TunnelClientImpl) readResponses() {
	scanner := bufio.NewScanner(c.conn)
	for scanner.Scan() {
		msg := scanner.Text()
		var serverMsg ServerMessage
		if err := json.Unmarshal([]byte(msg), &serverMsg); err != nil {
			log.Printf("Error parsing server message: %v", err)
			continue
		}

		log.Printf("Received from tunnel server: %+v", serverMsg)

		// Handle specific message types
		if serverMsg.Action == "PONG" {
			log.Printf("Received PONG from tunnel server")
		}
	}

	if err := scanner.Err(); err != nil {
		log.Printf("Error reading from tunnel server: %v", err)
	}

	// Connection closed, attempt reconnect
	c.mutex.Lock()
	c.isConnected = false
	c.mutex.Unlock()

	c.attemptReconnect()
}

// attemptReconnect tries to reconnect to the tunnel server
func (c *TunnelClientImpl) attemptReconnect() {
	c.mutex.Lock()
	defer c.mutex.Unlock()

	// Stop the ping ticker
	if c.pingTicker != nil {
		c.pingTicker.Stop()
		c.pingTicker = nil
	}

	// Check if we should stop
	select {
	case <-c.stopChan:
		return
	default:
		// Continue with reconnect
	}

	c.reconnectCount++
	reconnectDelay := time.Duration(c.config.ReconnectDelay) * time.Second

	log.Printf("Connection to tunnel server lost. Attempting reconnect %d in %v...",
		c.reconnectCount, reconnectDelay)

	// Schedule reconnect
	time.AfterFunc(reconnectDelay, func() {
		err := c.Connect()
		if err != nil {
			log.Printf("Failed to reconnect to tunnel server: %v", err)
			c.attemptReconnect() // Try again
		}
	})
}

// startPingTicker starts a ticker to send periodic pings
func (c *TunnelClientImpl) startPingTicker() {
	c.mutex.Lock()
	defer c.mutex.Unlock()

	if c.pingTicker != nil {
		c.pingTicker.Stop()
	}

	pingInterval := time.Duration(c.config.PingInterval) * time.Second
	c.pingTicker = time.NewTicker(pingInterval)

	go func() {
		for {
			select {
			case <-c.pingTicker.C:
				c.sendPing()
			case <-c.stopChan:
				return
			}
		}
	}()
}

// sendPing sends a ping message to keep the connection alive
func (c *TunnelClientImpl) sendPing() {
	c.mutex.Lock()
	defer c.mutex.Unlock()

	if !c.isConnected || c.conn == nil {
		return
	}

	pingMsg := PingMessage{
		Action: "PING",
	}

	pingBytes, err := json.Marshal(pingMsg)
	if err != nil {
		log.Printf("Failed to marshal ping message: %v", err)
		return
	}

	_, err = c.conn.Write(append(pingBytes, '\n'))
	if err != nil {
		log.Printf("Failed to send ping: %v", err)
		// Connection might be broken, close it to trigger reconnect
		c.conn.Close()
		c.isConnected = false
		return
	}

	log.Printf("Sent PING to tunnel server")
}

// Disconnect closes the connection to the tunnel server
func (c *TunnelClientImpl) Disconnect() {
	c.mutex.Lock()
	defer c.mutex.Unlock()

	// Signal to stop reconnect attempts and ping ticker
	close(c.stopChan)

	if c.pingTicker != nil {
		c.pingTicker.Stop()
		c.pingTicker = nil
	}

	if c.conn != nil {
		c.conn.Close()
		c.conn = nil
	}

	c.isConnected = false
	log.Printf("Disconnected from tunnel server")
}

// IsConnected returns whether the client is currently connected
func (c *TunnelClientImpl) IsConnected() bool {
	c.mutex.Lock()
	defer c.mutex.Unlock()
	return c.isConnected
}

File: devtools/KNIRVSDK/ts/transmission/knirv-tunnel-client/tunnelClient.js
// knirv-tunnel-client/tunnelClient.js
const net = require('net');
const os = require('os');

class KNIRVTunnelClient {
    constructor(options) {
        this.tunnelServerHost = options.tunnelServerHost || process.env.TUNNEL_SERVER_HOST || ROOTCHAIN_URL;
        this.tunnelServerControlPort = options.tunnelServerControlPort || 4001;
        this.peerId = options.peerId; // Required
        this.chainId = options.chainId;
        this.internalIp = options.internalIp || this._getLocalIp();
        this.internalP2pPort = options.internalP2pPort || 5050;
        this.type = options.type || 'peer';
        
        this.controlSocket = null;
        this.isConnected = false;
        this.reconnectAttempts = 0;
        this.maxReconnectAttempts = options.maxReconnectAttempts || 10;
        this.reconnectDelay = options.reconnectDelay || 5000; // 5 seconds
    }
    
    _getLocalIp() {
        // Simple function to get a local IP address
        const interfaces = os.networkInterfaces();
        for (const name of Object.keys(interfaces)) {
            for (const iface of interfaces[name]) {
                if (iface.family === 'IPv4' && !iface.internal) {
                    return iface.address;
                }
            }
        }
        return '127.0.0.1'; // Fallback
    }
    
    connect() {
        if (this.isConnected) {
            console.log('Already connected to tunnel server.');
            return;
        }
        
        console.log(`Connecting to tunnel server at ${this.tunnelServerHost}:${this.tunnelServerControlPort}...`);
        
        this.controlSocket = new net.Socket();
        
        this.controlSocket.on('connect', () => {
            console.log('Connected to tunnel server. Sending identification...');
            this.isConnected = true;
            this.reconnectAttempts = 0;
            
            // Send identification message
            const identifyMsg = {
                action: 'IDENTIFY',
                peerId: this.peerId,
                chainId: this.chainId,
                internalIp: this.internalIp,
                internalP2pPort: this.internalP2pPort,
                type: this.type
            };
            
            this.controlSocket.write(JSON.stringify(identifyMsg) + '\n');
        });
        
        this.controlSocket.on('data', (data) => {
            try {
                const messages = data.toString().split('\n').filter(msg => msg.trim());
                for (const msg of messages) {
                    const parsed = JSON.parse(msg);
                    console.log('Received from tunnel server:', parsed);
                    
                    if (parsed.action === 'PONG') {
                        console.log('Received PONG from tunnel server.');
                    }
                }
            } catch (e) {
                console.error('Error parsing message from tunnel server:', e.message);
            }
        });
        
        this.controlSocket.on('error', (err) => {
            console.error('Control socket error:', err.message);
            this.isConnected = false;
        });
        
        this.controlSocket.on('close', () => {
            console.log('Control socket closed.');
            this.isConnected = false;
            
            // Attempt to reconnect
            if (this.reconnectAttempts < this.maxReconnectAttempts) {
                this.reconnectAttempts++;
                console.log(`Attempting to reconnect (${this.reconnectAttempts}/${this.maxReconnectAttempts}) in ${this.reconnectDelay}ms...`);
                setTimeout(() => this.connect(), this.reconnectDelay);
            } else {
                console.error('Max reconnect attempts reached. Giving up.');
            }
        });
        
        // Connect to the tunnel server
        this.controlSocket.connect(this.tunnelServerControlPort, this.tunnelServerHost);
    }
    
    disconnect() {
        if (this.controlSocket && !this.controlSocket.destroyed) {
            console.log('Disconnecting from tunnel server...');
            this.controlSocket.destroy();
        }
    }
    
    // Send a ping to keep the connection alive
    ping() {
        if (this.isConnected && this.controlSocket && !this.controlSocket.destroyed) {
            this.controlSocket.write(JSON.stringify({ action: 'PING' }) + '\n');
        }
    }
    
    // Start a ping interval
    startPingInterval(interval = 30000) {
        this._pingInterval = setInterval(() => this.ping(), interval);
    }
    
    // Stop the ping interval
    stopPingInterval() {
        if (this._pingInterval) {
            clearInterval(this._pingInterval);
            this._pingInterval = null;
        }
    }
}

module.exports = KNIRVTunnelClient;